### DISCUSSION TITLE
【萌新求助】如何以消息中的表情为关键词触发回复

### DISCUSSION THREAD

**Discussion by the-cold-heart** [nonebot/discussions#191] [问答] [answered]
如题，如何以消息中的表情（如：[CQ:face,id=178]）为关键词触发回复？

**Comment by yanyongyu** [comment, marked as answer]
你可以自己写一个rule判断消息中的表情，然后`on_message(rule)`即可

**Reply by the-cold-heart** [reply]
写rule我倒是知道，但是如何判断消息中的表情呢？我尝试过使用event.get_plaintext() == "[CQ:face,id=178]" ，但是并不奏效，即使是单独发送这个表情也不会被检测到QWQ

**Reply by yanyongyu** [reply]
```python
face_ids = [seg.data["id"] for seg in event.get_message()["face"]]
```

**Reply by the-cold-heart** [reply]
没太看懂该怎么用啊，能详细讲解一下吗QWQ

**Reply by yanyongyu** [reply]
就是提取消息中所有的表情id列表

**Reply by the-cold-heart** [reply]
请原谅我愚钝，我还是不知道该怎么做，可以写一个完整的示例来讲解吗QWQ

**Reply by yanyongyu** [reply]
如果你是要当消息只有一个表情触发：

```python
from nonebot.params import EventMessage
from nonebot.adapters.onebot.v11 import Message

async def checker(message: Message = EventMessage) -> bool:
	return message == Message(MessageSegment.face(178))
```

如果你是要消息中包含这个表情：

```python
from nonebot.params import EventMessage
from nonebot.adapters.onebot.v11 import Message

async def checker(message: Message = EventMessage) -> bool:
	return MessageSegment.face(178) in message
```

**Reply by the-cold-heart** [reply]
我试着写了一下，成功触发了，但是并没有回复我，不知道为什么，代码如下
```python
async def test_checker(message: Message = EventMessage) -> bool:
	return MessageSegment.face(178) in message
test = on_message(rule = test_checker)
@test.handle()
async def test_handle():
    await test.finish(Message("成功！"))
```
后台给我报了个这个
`[DEBUG] nonebot |  type <class 'function'> not match depends <function test_checker at 0x00000157ED735040> annotation Message, ignored`

**Reply by yanyongyu** [reply]
dbq少打了个括号

```diff
- async def checker(message: Message = EventMessage) -> bool:
+ async def checker(message: Message = EventMessage()) -> bool:
```

**Reply by the-cold-heart** [reply]
成功了！不过给后来的人提个醒，示例中没有定义MessageSegment，记得在代码开头加上
```python
from nonebot.adapters.onebot.v11 import MessageSegment
```
最后，**感谢大佬指导！！！**

**Reply by zhulinyv** [reply]
有没有办法能检测同一表情的数量然后回复不同信息？我想用re正则表达式筛选，可是不知道该如何写。

**Reply by yanyongyu** [reply]
```python
face_ids = [seg.data["id"] for seg in event.get_message()["face"]]
face_ids.count(178)
```

**Reply by zhulinyv** [reply]
> ```python
> face_ids = [seg.data["id"] for seg in event.get_message()["face"]]
> face_ids.count(178)
> ```

抱歉，有点不知道怎么用，知道数量之后怎么回复不同的信息呢T_T

**Reply by yanyongyu** [reply]
你自己if elif else判断啊。。。

**Reply by zhulinyv** [reply]
> 你自己if elif else判断啊。。。

判断过了，可是我不知道怎么让它返回特定的值。。。
![image](https://user-images.githubusercontent.com/66541860/183282001-ba686163-d638-44f4-a725-ea6f6f490d58.png)
![$ @F8CR~@2P5R5E6Y~6_E$H](https://user-images.githubusercontent.com/66541860/183282067-3405770a-e122-4076-bc6e-6ffdbc9f73ee.jpg)

**Reply by yanyongyu** [reply]
不要把checker和判断数量写在一起啊，另外写一个函数，自己传参数，或者用Depends

**Reply by the-cold-heart** [reply]
我想……也许你可以创建多个rule和多个事件响应器，一个rule对应一个事件响应器，其中rule可以设置第一个为等于一的，第二个设置为等于二的，第三个可以设置为大于二的，然后一一对应上事件响应器就好了（我也是个菜鸟，只是提供一种思路，表达的不是很清楚，见谅）

**Reply by the-cold-heart** [reply]
如果一定要写在一起的话，可以在rule里面先判断消息中是否有这个表情，如果有，加一个全局变量，变量为int型，其值等于表情的数量，并返回值为True，然后在事件响应器中使用if，elif，else通过刚才创建的全局变量来判断表情数量，并做出不同的回复

**Reply by yanyongyu** [reply]
直接把表情计数写在事件处理里就行了，不用动rule

**Reply by zhulinyv** [reply]
好的，了解了，谢谢大佬们

**Reply by zhulinyv** [reply]
我又来打扰大佬们了🥲这样写为什么报错，我为什么不能直接调用它😭
![image](https://user-images.githubusercontent.com/66541860/183286144-f245dbf3-86c3-4be8-8605-51418bb6288b.png)

![image](https://user-images.githubusercontent.com/66541860/183286101-b7f9780c-f05c-42a2-8588-918d021e292e.png)

**Reply by the-cold-heart** [reply]
试试这个吧
```python
async def candy_checker(message: Message = EventMessage()) -> bool:
	return MessageSegment.face(147) in message
candy = on_message(rule = candy_checker)
@candy.handle()
async def candy_handle():
	face_ids = [seg.data["id"] for seg in event.get_message()["face"]]
	if face_ids.count(147) == 1:
		await candy.finish(Message("回复一"))
	elif face_ids.count(147) == 2:
		await candy.finish(Message("回复二"))
	else:
		await candy.finish(Message("回复三"))
```

**Reply by zhulinyv** [reply]
> 

谢谢大佬啦，不过不知道为什么，不管发送几个棒棒糖，都是回复三😭

**Reply by the-cold-heart** [reply]
那应该是计数出了问题，说实话，我也不太清楚怎么获取相关表情的数量，但是解决了表情计数应该就可以成功了，至于如何计数还是等大佬来讲吧……

**Reply by zhulinyv** [reply]
不过还是谢谢啦，果然我还是适合用大佬们写好的插件😢

**Reply by yanyongyu** [reply]
> 谢谢大佬啦，不过不知道为什么，不管发送几个棒棒糖，都是回复三sob

应该是count("147")

**Reply by zhulinyv** [reply]
> > 谢谢大佬啦，不过不知道为什么，不管发送几个棒棒糖，都是回复三sob
> 
> 应该是count("147")

谢谢佬，解决了╰(*°▽°*)╯开心

**Comment by 17TheWord** [comment]
```python
# 获取消息
get_msg = event.raw_message

# 判断消息中是否包含表情的CQ码
if "[CQ:face,id=178]" in get_msg:
    # 结果为True
    # 执行代码
else:
    # 结果为False
    # 执行代码
```
